include/fastscapelib/sinks.hpp

```diff
@@ -33,98 +34,95 @@ namespace fastscapelib
          * resolved.
          */
-        template<class E, class elev_t = typename std::decay_t<E>::value_type>
-        void init_pflood(E&& elevation,
-                        xt::xtensor<bool, 2>& closed,
-                        node_pr_queue<elev_t>& open)
+        template<class G, class E, class elev_t = typename std::decay_t<E>::value_type>
+        void init_pflood(G& grid,
+                         E&& elevation,
+                         xt::xtensor<bool, 1>& closed,
+                         node_pr_queue<G, elev_t>& open)
         {
-            auto elev_shape = elevation.shape();
+            using size_type = typename G::size_type;
 
-            index_t nrows = static_cast<index_t>(elev_shape[0]);
-            index_t ncols = static_cast<index_t>(elev_shape[1]);
-
-            auto place_node = [&](index_t row, index_t col)
-            {
-                open.emplace(node_container<elev_t>(row, col, elevation(row, col)));
-                closed(row, col) = true;
-            };
+            // TODO: assert elevation shape match grid shape
 
-            for(index_t c=0; c<ncols; ++c)
-            {
-                place_node(0, c);
-                place_node(nrows-1, c);
-            }
+            const auto elevation_flat = xt::flatten(elevation);
 
-            for(index_t r=1; r<nrows-1; ++r)
+            for (size_type idx=0; idx<grid.size(); ++idx)
             {
-                place_node(r, 0);
-                place_node(r, ncols-1);
+                if (grid.status_at_nodes()[idx] == fastscapelib::node_status::fixed_value_boundary)
+                {
+                    open.emplace(node_container<G, elev_t>(idx, elevation_flat(idx)));
+                    closed(idx) = true;
+                }
             }
         }
 
 
         /**
          * fill_sinks_flat implementation.
          */
-        template<class E>
-        void fill_sinks_flat_impl(E&& elevation)
+        template<class G, class E>
+        void fill_sinks_flat_impl(G& grid, E&& elevation)
         {
+            using size_type = typename G::size_type;
             using elev_t = typename std::decay_t<E>::value_type;
-            auto elev_shape = elevation.shape();
 
-            node_pr_queue<elev_t> open;
-            xt::xtensor<bool, 2> closed = xt::zeros<bool>(elev_shape);
+            auto elevation_flat = xt::flatten(elevation);
 
-            init_pflood(elevation, closed, open);
+            node_pr_queue<G, elev_t> open;
+            xt::xtensor<bool, 1> closed = xt::zeros<bool>({grid.size()});
+
+            init_pflood(grid, elevation, closed, open);
 
             while(open.size()>0)
             {
-                node_container<elev_t> inode = open.top();
+                node_container<G, elev_t> inode = open.top();
                 open.pop();
 
-                for(unsigned short k=1; k<=8; ++k)
+                for(auto nidx : grid.neighbors_indices(inode.m_idx))
```

Comment:
I think it could be more efficient to:
```cpp
neighbors_type neighbors;  // create a neighbors container to fill
grid.neighbors(inode.m_idx, neighbors);  // fill it
for (auto n=neighbors.begin(); n != neighbors.end(); ++n)
{          
// do some job
}
```
like it's implemented [here](https://github.com/fastscape-lem/fastscapelib/blob/b19a051fc02a1e19105288f7ba766ec44aad0840/include/fastscapelib/flow_routing.hpp#L315)

That way you always reuse the same container and don't pay allocation at each call.
Maybe you could think yo something similar for `indices` method.

Reply:
Ah yes I see.

> Maybe we could think to something similar for indices method.

I'll open another issue for this.

Reply:
> I'll open another issue for this.

Opened #71. Please let me know what you think about that.